src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,85 +1081,165 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
+                const auto ret{sisl::Compress::decompress(r_cast< const char* >(buf->bytes),
+                                                          r_cast< char* >(decompressed_buf->bytes),
+                                                          mblk->hdr.h.compressed_sz, &decompressed_size)};
+                if (ret != 0) {
+                    LOGERROR("[type={}], negative result: {} from decompress trying to decompress the "
+                             "data. compressed_sz: {}, src_context_sz: {}",
+                             mblk->hdr.h.type, ret, uint64_cast(mblk->hdr.h.compressed_sz),
+                             uint64_cast(mblk->hdr.h.src_context_sz));
+                    HS_REL_ASSERT(false, "failed to decompress");
                 } else {
-                    // There is use case that cb could be nullptr because client want to get its superblock via
-                    // read api;
-                    cb(mblk, buf, mblk->hdr.h.context_sz);
+                    // decompressed_size must equal to input sz before compress
+                    HS_REL_ASSERT_EQ(uint64_cast(mblk->hdr.h.src_context_sz),
+                                     uint64_cast(decompressed_size)); /* since decompressed_size is >=0 it
+                                                                         is safe to cast to uint64_t */
+                    HS_LOG(DEBUG, metablk,
+                           "[type={}] Successfully decompressed, compressed_sz: {}, src_context_sz: {}, "
+                           "decompressed_size: {}",
+                           mblk->hdr.h.type, uint64_cast(mblk->hdr.h.compressed_sz),
+                           uint64_cast(mblk->hdr.h.src_context_sz), decompressed_size);
                 }
 
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
-                       uint64_cast(mblk->hdr.h.context_sz));
+                cb(mblk, decompressed_buf, mblk->hdr.h.src_context_sz);
+            } else {
+                // There is use case that cb could be nullptr because client want to get its superblock via
+                // read api;
+                cb(mblk, buf, mblk->hdr.h.context_sz);
             }
-        } else {
-            HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
+
+            HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
+                   uint64_cast(mblk->hdr.h.context_sz));
+        }
+    } else {
+        HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
 #if 0
             // should never arrive here since we do assert on type before write to disk;
             HS_LOG_ASSERT( false, "[type={}] not registered for mblk found on disk. Skip this meta blk. ",
                       mblk->hdr.h.type);
 #endif
+    }
+}
+
+std::vector< meta_sub_type >
+MetaBlkService::topological_sort(std::unordered_map< meta_sub_type, std::vector< meta_sub_type > >& topological_graph) {
```

Comment:
Move this to `class hs_utils`, replace meta_sub_type to std::string, so that it can be reused as homestore utility.